tensorflow/python/keras/engine/training.py

```diff
@@ -1084,27 +1084,41 @@ def fit(self,
       # Handle fault-tolerance for multi-worker.
       # TODO(omalleyt): Fix the ordering issues that mean this has to
       # happen after `callbacks.on_train_begin`.
-      data_handler._initial_epoch = (  # pylint: disable=protected-access
-          self._maybe_load_initial_epoch_from_ckpt(initial_epoch))
-      logs = None
+      initial_epoch = self._maybe_load_initial_epoch_from_ckpt(initial_epoch)
+      data_handler._initial_epoch = initial_epoch  # pylint: disable=protected-access
+      run_compiled = (not self.run_eagerly and
+                      not callbacks._should_call_train_batch_hooks and  # pylint: disable=protected-access
+                      self._steps_per_execution.numpy().item() == 1)
+      logs, epoch_train_fn = None, None
       for epoch, iterator in data_handler.enumerate_epochs():
         self.reset_metrics()
         callbacks.on_epoch_begin(epoch)
-        with data_handler.catch_stop_iteration():
-          for step in data_handler.steps():
-            with trace.Trace(
-                'train',
-                epoch_num=epoch,
-                step_num=step,
-                batch_size=batch_size,
-                _r=1):
-              callbacks.on_train_batch_begin(step)
-              tmp_logs = self.train_function(iterator)
-              if data_handler.should_sync:
-                context.async_wait()
-              logs = tmp_logs  # No error, now safe to assign to logs.
-              end_step = step + data_handler.step_increment
-              callbacks.on_train_batch_end(end_step, logs)
+        if (run_compiled and data_handler.inferred_steps is not None and
+            epoch - initial_epoch > 0):
+          if epoch_train_fn is None:
+            @def_function.function(experimental_relax_shapes=True)
+            def epoch_train_fn(iterator, steps, logs):
+              for _ in math_ops.range(steps):
+                logs = self.train_function(iterator)
+              return logs
+
+          logs = epoch_train_fn(iterator, data_handler.inferred_steps, logs)
+        else:
+          with data_handler.catch_stop_iteration():
+            for step in data_handler.steps():
+              with trace.Trace(
+                  'train',
+                  epoch_num=epoch,
+                  step_num=step,
+                  batch_size=batch_size,
+                  _r=1):
```

Comment:
I didn't add tracing to the compiled code path as the `TensorBoard` profiling callback needs to have batch level hooks defined anyway so it won't be used in the compiled version anyway as far as I know.